proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,506 @@
+identifying the source of such errors. This leads to a poor debugging experience
+and makes it difficult to monitor systems in production.
+
+While over the past few years GHC has grown a variety of mechanisms for reporting
+backtraces (e.g. ``HasCallStack``, ``GHC.Stack.CCS``, and DWARF debug
+information), currently we do not have a means to attach such backtraces to
+exceptions. The goal of this proposal is to fix this `long-standing
+<https://www.youtube.com/watch?v=J0c4L-AURDQ>`_ problem.
+
+We want to ensure that exceptions report provenance information *by
+default* without requiring action on the part of the developer. To provide this provenance we leverage
+the existing mechanisms for collecting backtraces listed above. Furthermore, we
+want to ensure that this information is available for consumption in structured
+form by the user program, to allow use by logging libraries (for instance, see
+`kaptip-raven #1
+<https://github.com/cachix/katip-raven/issues/1#issuecomment-625389463>`_),
+automatic error reporting, code analysis tools, and the like.
+
+Proposed Change Specification
+-----------------------------
+
+This proposal consists of three largely-independent parts:
+
+1. augmenting the current root of the exception hierarchy,
+   ``SomeException``, with additional metadata in the form of an
+   ``ExceptionContext``.
+2. the introduction of annotation types to capture backtraces from the
+   above-described backtrace-collection mechanisms
+3. a facility for choosing which backtrace mechanism(s)
+   should be used to collect exception provenance.
+
+We will describe these parts below.
+
+Introducing ``ExceptionContext``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+All exceptions in Haskell are currently represented by the
+``Control.Exception.SomeException`` type which captures the exception along
+with its ``Exception`` dictionary: ::
+
+    data SomeException where
+        SomeException :: forall a. (Exception a)
+                      => a -> SomeException
+
+Such a ``SomeException`` value carries all of the information needed to catch,
+identify, and display an exception, but lacks any generic facility for encoding
+relevant metadata about the exception.
+
+As a first step towards attaching backtraces to exceptions, we propose to extend
+``SomeException`` to carry an additional "context" value, consisting of a collection
+of dynamically-typed annotations. The dynamic nature of these annotations
+allows the mechanism to handle not just backtraces but any arbitrary
+context-relevant value which may be inspected by a handler for analysis and logging.
+To avoid breaking existing users of ``SomeException``, we introduce this
+context as an implicit parameter constraint:  ::
+
+    data SomeException where
+        SomeException :: forall a. (Exception a, ?context :: ExceptionContext)
+                      => a -> SomeException
+
+    data ExceptionContext = ExceptionContext [SomeExceptionAnnotation]
```

Comment:
A call stack is a single annotation. Until now I had been considered `ExceptionContext` to be an unordered bag of annotations and indeed through that lens the ordered natured of list is a bit odd. One way of addressing this would be to state that `ExceptionContext`s are to be interpreted as *stacks* of annotations (which would roughly reflect the lexical scoping of the handlers which added the annotations). This isn't *strictly* true since multiple annotations may be added by the same call to, e.g., `withExceptionContext`. However, I think it's true enough to be a useful convention and at least somewhat resolve the semantic issue that you raise.

 * 


 I do agree that semantically the ordered nature of the list representation is somewhat odd as `ExceptionContext` is really an unordered bag of annotations. Note that it's not necessarily a map; there may be more than one `HasCallStackBacktrace` in the However, practically-speaking list is one of the few structures which we can readily use in `base`; I am not sure that "somewhat odd" really justifies adding a map implementation